lib/SimplifyPointerBitcastPass.cpp

```diff
@@ -275,8 +279,19 @@ bool clspv::SimplifyPointerBitcastPass::runOnGEPFromGEP(Module &M) const {
     SmallVector<Value *, 8> Idxs;
 
     Value *SrcLastIdxOp = OtherGEP->getOperand(OtherGEP->getNumOperands() - 1);
-
     Value *GEPIdxOp = GEP->getOperand(1);
+
+    if (clspv::PointersAre64Bit(M) && OtherGEP->getType()->isStructTy()) {
```

Comment:
Another instance I wonder why we aren't just working in the index type throughout. Am I missing an assumption why we would never overflow?

Reply:
We don't control all the GEPs coming into the pass, so quite frequently there are a mix of i32 and i64 indices. Even with the spir64 target, small constant indices are usually i32. So here the pass will:

- Promote all i32 indices to i64. This means we can do arithmetic on the different indices in a GEP without the types mismatching.
- At the end, demote i64 indices back down i32 for struct member indices, since LLVM requires this. There is a check for constant values that they still fit in an i32, but in theory this is only a problem for structs that have more than UINT_MAX members.

I'm reasonably confident that this is ok, but definitely open to other ideas of what to do here.

Reply:
Thinking about this a bit more, the type being a StructTy doesn't guarantee that the current GEP index operand being looked at is the actual index to a struct member, it could be an index into a buffer of structs, which might then overflow when truncated. I think this needs to be more precise to only truncate certain indices when it find a struct, rather than all of them.

Reply:
It turns out this entire block of code was redundant. I've tidied it up now so the only new changes are:

- Extending indices to i64 when needed to make arithmetic work
- Truncating stray i64 struct member indices back down to i32, but only when we actually know the constant value will fit